### PR TITLE
Reapply cg_llvm: `fewer_names` in `uncached_llvm_type`

### PR DESCRIPTION
**Pull request by tmiasko**
r? @davidtwco @erikdesjardins

### CODE REVIEW

**Review by davidtwco — APPROVED** [review]
Thanks for reviving this. You don't need to, but I'd appreciate it if you added myself and @erikdesjardins as co-authors to the commit.

**Comment by marmeladema** [comment]
I rerun my tests and the problem appears indeed to have been fixed:
Against stable
```
$ cargo run --release
    Finished release [optimized + debuginfo] target(s) in 0.29s
     Running `target/release/bitvec-perf-regression`
instructions:u = 871040161
```
Against f12f59345fd5621cd34aea2915fff165082cd5d0
```
$ cargo +f12f59345fd5621cd34aea2915fff165082cd5d0 run --release
    Finished release [optimized + debuginfo] target(s) in 0.01s
     Running `target/release/bitvec-perf-regression`
instructions:u = 870400148
```

Since the problem has been fixed by LLVM 14 should this change be gated by the LLVM version or postponed until LLVM 14 becomes the oldest supported version? Otherwise people compiling rustc manually with LLVM version < 14 will hit the performance regression right?

File: compiler/rustc_codegen_llvm/src/type_of.rs [hunk 1]
```diff
@@ -42,7 +43,12 @@ fn uncached_llvm_type<'a, 'tcx>(
         // FIXME(eddyb) producing readable type names for trait objects can result
         // in problematically distinct types due to HRTB and subtyping (see #47638).
         // ty::Dynamic(..) |
-        ty::Adt(..) | ty::Closure(..) | ty::Foreign(..) | ty::Generator(..) | ty::Str => {
+        ty::Adt(..) | ty::Closure(..) | ty::Foreign(..) | ty::Generator(..) | ty::Str
+            // For performance reason we use names only when emitting LLVM IR. Unless we are on
```
Comment (nikic) [inline compiler/rustc_codegen_llvm/src/type_of.rs:47]:
```suggestion
            // For performance reasons we use names only when emitting LLVM IR. Unless we are on
```

File: compiler/rustc_codegen_llvm/src/type_of.rs [hunk 2]
```diff
@@ -42,7 +43,12 @@ fn uncached_llvm_type<'a, 'tcx>(
...
+            // LLVM < 14, where the use of unnamed types where involved in various issues, e.g.,
```
Comment (nikic) [inline compiler/rustc_codegen_llvm/src/type_of.rs:48]:
```suggestion
            // LLVM < 14, where the use of unnamed types resulted in various issues, e.g.,
```
or something, not sure what "where involved" was supposed to be here.

**Comment by tmiasko** [comment]
* Gated the use of unnamed types on LLVM 14 version check.
* Dropped test changes src/test/ui/const-generics/issues/issue-75763.rs. This pull request merely hides the bug, and I plan to fix the underlying issue separately.

I would rather avoid including this in next beta, and if I am reading Rust forge correctly, the next promotion is on Monday, so I am marking this as blocked at least until then.

**Comment by davidtwco** [comment]
r=me when you're ready to land this

**Pull request merged**